Switch <Context> to mean <Context.Provider>

gaearon opened the pull request:
Previously, `<Context>` was equivalent to `<Context.Consumer>`. However, since the introduction of Hooks, the `<Context.Consumer>` API is rarely used. The goal here is to make the common case cleaner:

```js
const ThemeContext = createContext('light')

function App() {
  return (
    <ThemeContext value="dark">
      ...
    </ThemeContext>
  )
}

function Button() {
  const theme = use(ThemeContext)
  // ...
}
```

This is technically a breaking change, but we've been warning about rendering `<Context>` directly for several years by now, so it's unlikely much code in the wild depends on the old behavior. [Proof that it warns today (check console).](https://codesandbox.io/p/sandbox/peaceful-nobel-pdxtfl)

---

**The relevant commit is 5696782b428a5ace96e66c1857e13249b6c07958.** It switches `createContext` implementation so that `Context.Provider === Context`.

The main assumption that changed is that a Provider's fiber type is now the context itself (rather than an intermediate object). Whereas a Consumer's fiber type is now always an intermediate object (rather than it being sometimes the context itself and sometimes an intermediate object).

My methodology was to start with the relevant symbols, work tags, and types, and work my way backwards to all usages.

This might break tooling that depends on inspecting React's internal fields. I've added DevTools support in the second commit. This didn't need explicit versioning—the structure tells us enough.

gaearon commented on packages/shared/ReactSymbols.js:
```diff
@@ -17,7 +17,7 @@ export const REACT_PORTAL_TYPE: symbol = Symbol.for('react.portal');
 export const REACT_PROFILER_TYPE: symbol = Symbol.for('react.profiler');
-export const REACT_PROVIDER_TYPE: symbol = Symbol.for('react.provider');
+export const REACT_CONSUMER_TYPE: symbol = Symbol.for('react.consumer');
 export const REACT_CONTEXT_TYPE: symbol = Symbol.for('react.context');
```
This is the crux of the change.

Providers shifted from `REACT_PROVIDER_TYPE` (deleted) to `REACT_CONTEXT_TYPE`.

Consumers shifted from `REACT_CONTEXT_TYPE` to `REACT_CONSUMER_TYPE` (new).

This lets us avoid changing anything about `use(Foo)` or `contextType = Foo` and only change how we map a Fiber type to the underlying context.

gnoff commented on packages/shared/getComponentNameFromType.js:
```diff
@@ -100,10 +100,10 @@ export default function getComponentNameFromType(type: mixed): string | null {
-      case REACT_PROVIDER_TYPE:
-        const provider: ReactProviderType<any> = (type: any);
-        return getContextName(provider._context) + '.Provider';
+        return getContextName(context) + '.Provider';
```
I imagine we'll be in a world soon where you never think about `.Provider` and just assume rendering a context provides values and "using" a context reads values. In this context does it make sense for the component name to carry forward this notion of `.Provider`? or should we switch it and call `FooContext.Provider` `FooContext` instead of calling `FooContext` `FooContext.Provider`?

gaearon replied:
Hmm yea, I'd be happy to drop but figured it might invalidate snapshot tests or stuff like this. Maybe as a separate change later if this one settles?

gnoff reviewed (approved):
Let's hold on merging til we confirm we can but excited for this change

hoxyq commented on packages/react-debug-tools/src/ReactDebugHooks.js:
```diff
@@ -797,8 +796,7 @@ function setupContexts(contextMap: Map<ReactContext<any>, any>, fiber: Fiber) {
     if (current.tag === ContextProvider) {
-      const providerType: ReactProviderType<any> = current.type;
-      const context: ReactContext<any> = providerType._context;
+      const context: ReactContext<any> = current.type;
```
Would this work with previous versions of React, debugged by fresh build (with these changes) of React DevTools?

Should be keep both variants (`current.type?.context || current.type` for backwards-compatibility?

gaearon replied:
Hmm yeah, I forgot this is supposed to work for multiple versions.

gaearon commented on packages/react-debug-tools/src/ReactDebugHooks.js:
```diff
@@ -822,8 +821,11 @@ function setupContexts(contextMap: Map<ReactContext<any>, any>, fiber: Fiber) {
+      if ((context: any)._context !== undefined) {
+        // Support inspection of pre-19+ providers.
+        context = (context: any)._context;
+      }
```
Note: this isn't feature-flagged because it needs to support both.

gaearon commented on packages/react-server/src/ReactFizzServer.js:
```diff
@@ -1754,10 +1730,9 @@ function renderContextProvider(
+  context: ReactContext<any>,
   props: Object,
 ): void {
-  const context = type._context;
```
Moved this to the callsite.

gaearon commented on packages/react-server/src/ReactFizzServer.js:
```diff
@@ -1703,31 +1704,6 @@ function renderContextConsumer(
   context: ReactContext<any>,
   props: Object,
 ): void {
-  // The logic below for Context differs depending on PROD or DEV mode. In
```
Moved this to the callsite.

gaearon commented:
Rebased and added a feature flag. I didn't keep the old warnings but I did preserve the codepaths and data structures to make it safer to land. The feature flag is hardcoded `false` everywhere except `www` where it's dynamic.
